### ISSUE TITLE
Add Cast<T> and CastFrom<TDerived> to ImmutableArray<T>

### ISSUE THREAD

**Issue by jaredpar** [dotnet/corefx#400]
The `ImmutableArray<T>` type should have easier to find and use APIs for converting between related array element types.  

## Rationale and Usage
The `ImmutableArray<T>` type is meant to be as close as possible to `T[]` in functionality without the possibility of mutation.  One area where it is lacking or at least harder to discover is type conversions.   

With arrays it is possible to convert `T[]` to `U[]` when `T` inherits from `U`.  This conversion is implicit and requires no dynamic type checking. 

``` csharp
Dog[] dogArray = new Dog[] { } ;
Animal[] animalArray = dogArray;
```

The equivalent code for `ImmutableArray<T>` is less intuitive: 

``` csharp
ImmutableArray<Dog> dogArray = ImmutableArray.Create<Dog>();
ImmutableArray<Animal> animalArray = ImmutableArray.Create<Animal, Dog>(dogArray); 
```

There are a few issues with this approach:

- The API requires two type parameters which must be manually specified.  There is no possibility type inference will ever successfully infer them. 
- The API is not very discoverable.  The developer is likely looking for methods that have familiar conversion names (`As`, `Cast`, etc ...).  They will not stumble upon this particular method unless they go through every overload of `Create` and read the documentation. 
- The API here is named `Create` which implies an allocation.  All other `Immutable.Create` methods indeed allocate.  In this case though no allocation occurs, it simply repackages the underlying array in a new struct. 

Additonally the API is lacking in functionality.  It has a static up cast member (`ImmutableArary.Create<T, TDerived>`), a dynamic type check member (`ImmutableArray<T>.As`) but no dynamic down cast member.   Normal array code like the following has no direct mapping in `ImmutableArray<T>`

``` csharp
object[] array = GetValue();
Dog[] dogArray = (Dog[])array;
```

## Proposed API 
The first part of the proposal is to delete `ImmutableArray.Create<T, TDerived>`.  The method name does not line up with its functionality and placing it on `ImmutableArray` makes it impossible to take advantage of type inference.  

The following methods will be added to `ImmutableArray<T>`:

``` csharp
static ImmutableArray<T> CastFrom<TDerived>(ImmutableArray<TDerived> array)
    where TDerived : class, T;
ImmutableArray<TDerived> Cast<TDerived>();
```

Putting the methods here both increases their discoverability and allows us to take advantage of type inference.  The `T` value is always known hence the developer only has to specify the target type (as they do with standard arrays).  The original sample above becomes:

``` csharp
ImmutableArray<Dog> dogArray = ImmutableArray.Create<Dog>();
ImmutableArray<Animal> animalArray = ImmutableArray<Animal>.CastFrom(dogArray);
```

## Details

`CastFrom` will have the same functionality as `ImmutableArray::Create`.  It will be non-allocating and not generate any runtime type checks.  The constraints on the method are sufficient for the CLR to guarantee the conversion will succeed.

`Cast` will be non-allocating but will have a dynamic type check.  If that type check fails an exception will be generated (in the same manner as a normal C# array conversion).

**Comment by terrajobst** [comment]
Excellent write-up. That makes total sense to me. +1

**Comment by AArnott** [comment]
:+1: 
Although you have one sentence that seems misleading:

> It is possible to convert T[] to U[] in most cases where all elements in the array can convert to U.

The conversion's success has nothing to do with the types of elements in the array, and everything to do with how the array itself was created. A `T[]` can only convert to `U[]` if in fact `T[]` _already is_ a `U[]` that was previously upcast.

**Comment by jaredpar** [comment]
@AArnott 

I spent probably a good five minutes on that sentence and couldn't find anything I was truly happy with (that didn't expand into a mini version of the C# spec).

**Comment by weshaggard** [comment]
In general I like the idea and I support the functionality but I'm a little interested to see what others think about the names. We don't have any other methods in the framework called CastUp or CastDown. Perhaps we could call CastUp just Cast and CastDown CastDynamic or something along those lines.

**Comment by terrajobst** [comment]
I don't think thee are names we could recycle. `Cast` is wrong in my mind because Linq already provides a method with this name but it's not restricted to being an up-cast -- it will simply perform a direct cast.

I actually like the proposed names.

**Comment by kdelorey** [comment]
`CastUp` and `CastDown` make sense to me. I'm in favor if the proposed names.

**Comment by jaredpar** [comment]
I made an update to the proposal.  @stephentoub pointed out that the way I had `CastUp` defined wouldn't work because it required additional constraints on the class type parameter `T` from within a method. In order to work the types involved need to be inverted

- Class type parameter needs to be the parent type.
- Method type parameter needs to be the derived type.

The method was changed to reflect this and rename to `CastFrom` as it seems to be a more appropriate name. 

For the second method I personally prefer `Cast` over `CastDown`.  The reason being that the method can cast up, down and even side ways (between interfaces).  The more appropriate name is just `Cast`.

**Comment by terrajobst** [comment]
That makes sense. I agree that in that `Cast` is the right name. Can you update the proposal then?

**Comment by jaredpar** [comment]
@terrajobst updated

**Comment by weshaggard** [comment]
As a developer trying to use these APIs how am I to figure out that `CastFrom` actually doesn't allocation without reading documentation? I have a feeling a lot of people will just simply call `Cast` unless they know what the other one provides. In a case where people call `Cast` directly with a type can be up cast will it detect this and do the non-allocating version? If so that seems to argue that we just have the `Cast` method and if not is there a reason?

Another question you are proposing we remove a public API what impact will that have? Have we shipped a stable version with that API?

**Comment by jaredpar** [comment]
@weshaggard the term cast is generally associated with a type conversion, not an allocation.  

Neither of these APIs is allocating, it is just doing a type conversion under the hood.  The difference between the methods is `CastFrom` doesn't require a dynamic type check but `Cast` does.  

@nguerrera or @AArnott would have to comment on whether or not we've shipped with that API public before.

**Comment by weshaggard** [comment]
@jaredpar sorry I miss understood the allocation part after re-reading your proposal I think I confused it with the comparison to the other Create methods which allocate. 

In your code example what happens if people do this:

    ImmutableArray<Dog> dogArray = ImmutableArray.Create<Dog>();
    ImmutableArray<Animal> animalArray = dogArray.Cast<Animal>();

That seems to me like the most natural thing for someone to do. If that isn't substantially worse than calling the static `CastFrom` then why not just have the one `Cast` method? The fewer concepts people need to figure out the better.

**Comment by jaredpar** [comment]
@weshaggard the code will end up producing the same result with an additional type check.  The reason for adding the extra concept is that this type check can be significant in tight loops. 

Roslyn for example regularly executes this exact pattern in the parser:

- Create an `ImmutableArray<ChildNode>`
- Convert to an `ImmutableArray<ParentNode>` 

Profiles showed that the type check in this conversion was taking up a significant portion of parse time.  Moving to the non-type checking version was a clear win (and one that we simply cannot regress on).  

The awkward consumption of the existing `ImmutableArray.Create` API is what prompted this request.  Roslyn ended up writing their own version which had the basic features called out in this request.  

- [Roslyn Static Cast](http://source.roslyn.codeplex.com/#Microsoft.CodeAnalysis/Collections/StaticCast.cs,4d1240e7bc00cef0,references)

**Comment by weshaggard** [comment]
Thanks for the explanation I'm good with the additions. I would like to get clarification from @nguerrera and @AArnott before we remove the one Create method however.

**Comment by ellismg** [comment]
In the previous API Review meeting, I think it was mentioned that we had not shipped a stable version of ImmutableArray yet, and were free to make these changes.  Note that we just renamed ReverseContents to Reverse with #399, so we are already making changes along these lines.

**Comment by AArnott** [comment]
ImmutableArray itself has never shipped in a stable package, so we're good
there. And I agree with Jared's other comments.

**Comment by nguerrera** [comment]
Yes, we can still make breaking changes to ImmutableArray, but we had said that we'd keep them to a minimum if at all possible. I do see a few uses of `Create<T, TDerived>`, so we'd need to coordinate carefully with our partners.

**Comment by davkean** [comment]
One comment, the type parameter for Cast<TDerived> should be probably just Cast<T>, to be consistent with Linq, and to indicate that it can go both up and down.

**Comment by AArnott** [comment]
Won't Cast<T> hide the lynq extension method? Their behaviors, return types
and cases where they fail are different, so eclipsing one with the other
may be a bad idea.

**Comment by davkean** [comment]
Return type differences are fine - it's common to provide strongly typed versions of enumerable extension. How will their behaviors differ?

**Comment by jaredpar** [comment]
@davkean the LINQ `Cast` method is casting elements and the `ImmutableArray<T>` version is casting arrays.  Concrete example of where this would differ:

``` csharp
Animal a = new Dog(); 
var array = ImmutableArray.Create(a);  // Creates a Animal[] 
Enumerable.Cast<Dog>(array);  // Successfully returns IEnumerable<Dog> 
array.Cast<Dog>();  // Fail: can't convert Animal[] -> Dog[] 
```

**Comment by nguerrera** [comment]
@jaredpar  @davkean @AArnott Given that, I'm inclined to think that we should avoid hiding Enumerable.Cast with something slightly different.

**Comment by justinvp** [comment]
> Given that, I'm inclined to think that we should avoid hiding Enumerable.Cast with something slightly different.

`CastTo`?

**Comment by AArnott** [comment]
`CastTo` doesn't convey the significant difference IMO.
What about `CastArray`?

**Comment by jaredpar** [comment]
@AArnott `CastArray` is what I've been thinking about as well.

**Comment by davkean** [comment]
What's the behavior when the cast doesn't succeed? Can you call this out in the spec?

**Comment by jaredpar** [comment]
@davkean from the spec

>  Cast will be non-allocating but will have a dynamic type check. **If that type check fails an exception will be generated (in the same manner as a normal C# array conversion).**

Is there something else you wanted to say in addition to that?

**Comment by davkean** [comment]
Which exception, InvalidCastException?

**Comment by jaredpar** [comment]
@davkean it throws whatever the array conversion would throw.

**Comment by terrajobst** [comment]
[We've reviewed this proposal](https://github.com/dotnet/apireviews/blob/master/2015-01-14-misc/README.md#400-add-castt-and-castfromtderived-to-immutablearrayt) and we're fine this approach. We've some concerns around the naming but this can be reviewed on the PR.

**Comment by jaredpar** [comment]
Fixed via dbfb49bbdcef9346a60e9af4a0f8bc9e7126b22b
